serde/src/java/org/apache/hadoop/hive/serde2/JsonSerDe.java

```diff
@@ -127,7 +128,18 @@ public void initialize(Configuration conf, Properties tbl)
   public Object deserialize(Writable blob) throws SerDeException {
 
     Object row;
-    Text t = (Text) blob;
+    Text t;
+
+    /*To check writable instance and then create Text object
+    * In case of sequence file format which writes in key: org.apache.hadoop.io.LongWritable & value: org.apache.hadoop.io.BytesWritable
+    * format.
+    */
+    if (blob instanceof BytesWritable) {
+      t = new Text(((BytesWritable) blob).getBytes());
```

Comment:
i believe this will add an extra copy of the data. Can't you wrap the BytesWritable and produce the byte stream without a copy?

Reply:
Looking at the constructor of `Text` i can see that it is using `java.nio.charset.CharsetEncoder#encode(java.nio.CharBuffer)` that actually allocates a new buffer for every input (for every record in your case) 
```java
int n = (int)(in.remaining() * averageBytesPerChar());
        ByteBuffer out = ByteBuffer.allocate(n);
``` 
Thus it will be nice to avoid that extra overhead by wrapping the original byteWritable.

Reply:
Does, is it the right way to handle this ?

 @Override
  public Object deserialize(Writable blob) throws SerDeException {

    String txt = null;
    byte[] rowByteArr;

    try {
      if (blob instanceof Text) {
        txt = blob.toString().trim();
      } else if (blob instanceof BytesWritable) {
        rowByteArr = ((BytesWritable) blob).getBytes();
        txt = Text.decode(rowByteArr, 0, rowByteArr.length).trim();
      }

      assert txt != null;
      return structReader.parseStruct(new ByteArrayInputStream((txt.getBytes()), 0, txt.getBytes().length));
    } catch (Exception e) {
      LOG.warn("Error [{}] parsing json text [{}].", e, txt);
      throw new SerDeException(e);
    }
  }

Reply:
@b-slim : find changes in latest commits.

Reply:
i don't think it is the way to go, you are still encoding and decoding back the bytes.